File: Linked List/Arrays/81. Search in Rotated Sorted Array II/fighting.py
class Solution:
    def search(self, nums: list[int], target: int) -> bool:
        start = 0
        end = len(nums) - 1
        while start <= end:
            mid = (start + end) // 2
            if nums[mid] == target:
                return True

            if nums[start] < nums[mid]:
                if nums[start] <= target < nums[mid]:
                    end = mid - 1
                else:
                    start = mid + 1
            elif nums[start] == nums[mid]:
                start += 1
            else:
                if nums[mid] < target <= nums[end]:
                    start = mid + 1
                else:
                    end = mid - 1
        return False


temp = Solution()
nums = [2, 5, 6, 0, 0, 1, 2]
target = 7
print(temp.search(nums, target))

File: Linked List/Arrays/81. Search in Rotated Sorted Array II/fighting.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    bool search(vector<int> &nums, int target)
    {
        int start = 0;
        int end = nums.size() - 1;
        while (start <= end)
        {
            int mid = (start + end) / 2;

            if (nums[mid] == target)
            {
                return true;
            }

            if (nums[start] < nums[mid])
            {
                if (nums[start] <= target && nums[mid] > target)
                {
                    end = mid - 1;
                }
                else
                {
                    start = mid + 1;
                }
            }
            else if (nums[start] == nums[mid])
            {
                start++;
            }
            else
            {
                if (nums[mid] < target && target <= nums[end])
                {
                    start = mid + 1;
                }
                else
                {
                    end = mid - 1;
                }
            }
        }
        return false;
    }
};

int main()
{
    vector<int> nums;
    int mynums[] = {2, 5, 6, 0, 0, 1, 2};
    // nums.assign(mynums, mynums + sizeof(mynums) / sizeof(int));
    nums.assign(mynums, mynums + (*(&mynums + 1) - mynums));

    Solution temp;
    int target = 0;
    cout << temp.search(nums, target) << endl;
    // for (int i = 0; i < nums.size(); i++)
    // {
    //     cout << nums[i] << endl;
    // }
}